src/ring_theory/dedekind_domain/adic_valuation.lean

```diff
@@ -333,4 +333,72 @@ instance adic_completion.has_lift_t : has_lift_t K (v.adic_completion K) :=
 /-- The ring of integers of `adic_completion`. -/
 def adic_completion_integers : valuation_subring (v.adic_completion K) := valued.v.valuation_subring
 
+variables (R K)
+
+lemma adic_completion.is_integer (x : v.adic_completion K) :
+  x ∈ v.adic_completion_integers K ↔ (valued.v x : ℤₘ₀) ≤ 1 :=
+iff.rfl
+
+/-- The natural inclusion of `R` in `adic_completion_integers`, as a ring homomorphism. -/
+@[simps] def inj_adic_completion_integers : R →+* v.adic_completion_integers K :=
+{ to_fun := λ r, ⟨coe $ algebra_map R K r, by simpa only [adic_completion.is_integer,
+    valued.valued_completion_apply] using v.valuation_le_one _⟩,
+  map_one' := by simpa only [ring_hom.map_one],
+  map_mul' := λ x y,
+  begin
+    ext,
+    simp_rw [ring_hom.map_mul, subring.coe_mul, subtype.coe_mk, uniform_space.completion.coe_mul],
+  end,
+  map_zero' := by simpa only [ring_hom.map_zero],
+  map_add' := λ x y,
+  begin
+    ext,
+    simp_rw [ring_hom.map_add, subring.coe_add, subtype.coe_mk, uniform_space.completion.coe_add],
+  end, }
+
+lemma inj_adic_completion_integers.injective :
```

Comment:
This is equivalent to:
```lean
instance : no_zero_smul_divisors R (v.adic_completion_integers K) :=
```
right? I think this is likely to be a more useful statement.

Reply:
They are equivalent, yes, and I have changed it. Why do you think that this version is more useful? Is it just because `no_zero_smul_divisors` can be an instance?

Reply:
Yes, encoding this information using `no_zero_smul_divisors` will enable any lemmas which require this to apply. If we kept it as `lemma inj_adic_completion_integers.injective` then such lemmas could never work (unless we manually register it at point of use).